src/mpi/session/session_impl.c

```diff
@@ -76,45 +72,69 @@ int MPIR_Session_finalize_impl(MPIR_Session * session_ptr)
 int MPIR_Session_get_num_psets_impl(MPIR_Session * session_ptr, MPIR_Info * info_ptr,
                                     int *npset_names)
 {
-    int i = 0;
-    while (MPIR_pset_list[i]) {
-        i++;
-    }
-    *npset_names = i;
+    unsigned count =
+        MPIR_SESSION_NUM_DEFAULT_PSETS + utarray_len(MPIR_Process.pm_pset_array->parray);
+
+    /* Overflow check */
+    MPIR_Assert(count < INT_MAX);
+
+    *npset_names = (int) count;
     return MPI_SUCCESS;
 }
 
 int MPIR_Session_get_nth_pset_impl(MPIR_Session * session_ptr, MPIR_Info * info_ptr,
                                    int n, int *pset_len, char *pset_name)
 {
     int mpi_errno = MPI_SUCCESS;
+    char *uri = NULL;
+    bool uri_needs_free = false;
 
-    int i = 0;
-    while (MPIR_pset_list[i] && i < n) {
-        i++;
-    }
+    unsigned count =
+        MPIR_SESSION_NUM_DEFAULT_PSETS + utarray_len(MPIR_Process.pm_pset_array->parray);
 
-    if (!MPIR_pset_list[i]) {
+    /* Overflow check */
+    MPIR_Assert(count < INT_MAX);
+
+    if (n < 0 || n >= count) {
         MPIR_ERR_SETANDSTMT1(mpi_errno, MPI_ERR_ARG, goto fn_fail, "**psetinvalidn",
                              "**psetinvalidn %d", n);
+    } else if (n == MPIR_SESSION_WORLD_PSET_IDX) {
+        uri =
+            MPL_malloc((strlen(MPIR_SESSION_WORLD_PSET_NAME) + 1) * sizeof(char), MPL_MEM_SESSION);
+        MPL_strncpy(uri, MPIR_SESSION_WORLD_PSET_NAME, strlen(MPIR_SESSION_WORLD_PSET_NAME) + 1);
+        uri_needs_free = true;
+    } else if (n == MPIR_SESSION_SELF_PSET_IDX) {
+        uri = MPL_malloc((strlen(MPIR_SESSION_SELF_PSET_NAME) + 1) * sizeof(char), MPL_MEM_SESSION);
+        MPL_strncpy(uri, MPIR_SESSION_SELF_PSET_NAME, strlen(MPIR_SESSION_SELF_PSET_NAME) + 1);
+        uri_needs_free = true;
+    } else {
+        MPIR_Pset *pset;
+        mpi_errno =
+            MPIR_Pset_by_idx(MPIR_Process.pm_pset_array, n - MPIR_SESSION_NUM_DEFAULT_PSETS, &pset);
```

Comment:
I think each session need its own arrays of pointers to global psets.

Reply:
We discussed offline, that PMIX psets live globally and that each session accesses the global pset array for PMIx psets without the need for own pointers.